client/state/themes/actions/theme-transfer.js

```diff
@@ -154,11 +155,40 @@ function transferInitiateFailure( siteId, error, plugin, context ) {
  * @param {number} transferId -- the specific transfer
  * @param {number} [interval] -- time between poll attempts
  * @param {number} [timeout] -- time to wait for 'complete' status before bailing
+ * @param {boolean} hasThemeFileUpload -- has theme File Upload
  * @returns {Promise} for testing purposes only
  */
-export function pollThemeTransferStatus( siteId, transferId, interval = 3000, timeout = 180000 ) {
+export function pollThemeTransferStatus(
+	siteId,
+	transferId,
+	interval = 3000,
+	timeout = 240000,
+	hasThemeFileUpload = false
+) {
 	const endTime = Date.now() + timeout;
 	return ( dispatch ) => {
+		const tryThemeFetch = ( uploaded_theme_slug, resolve ) => {
+			wpcom.req.get( `/sites/${ siteId }/themes/mine`, { apiVersion: '1.1' } ).then( ( theme ) => {
+				if ( theme.id === uploaded_theme_slug ) {
+					dispatch( transferStatus( siteId, transferId, 'complete', '', uploaded_theme_slug ) );
+					dispatch( {
+						type: THEME_UPLOAD_SUCCESS,
+						siteId,
+						themeId: uploaded_theme_slug,
+					} );
+					// finished, stop polling
+					return resolve();
+				}
+
+				if ( Date.now() > endTime ) {
+					return setTimeout( tryThemeFetch, 1000, uploaded_theme_slug, resolve );
+				}
```

Comment:
Am I reading this wrong, or can this cause `tryThemeFetch` to run infinitely if the theme never gets activated after exceeding the allowed timeout?

Reply:
endTime should prevent that.